### ISSUE TITLE
test_pubsub_run_in_thread sometimes fails with RuntimeError: dictionary changed size during iteration

### ISSUE THREAD

**Issue by ivan** [jamesls/fakeredis#230]
I observed this with CPython 2.7.15 on Linux 4.14.83 on an Intel 4790K, just once so far:

```
======================================================================
ERROR: test_pubsub_run_in_thread (test_fakeredis.TestFakeStrictRedisDecodeResponses)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/fakeredis-0.16.0/test_fakeredis.py", line 3171, in test_pubsub_run_in_thread
    pubsub_thread.stop()
  File "/nix/store/q2svfj4gqvz77mq7sf0mm796kbic580q-python2.7-redis-2.10.6/lib/python2.7/site-packages/redis/client.py", line 2607, in stop
    self.pubsub.unsubscribe()
  File "/build/fakeredis-0.16.0/fakeredis.py", line 230, in decode_response
    val = _decode(func(*args, **kwargs))
  File "/build/fakeredis-0.16.0/fakeredis.py", line 508, in func_wrapper
    return func(*args, **kwargs)
  File "/build/fakeredis-0.16.0/fakeredis.py", line 2694, in unsubscribe
    *args)
  File "/build/fakeredis-0.16.0/fakeredis.py", line 2705, in _usubscribe
    for channel in subscribed_dict:
RuntimeError: dictionary changed size during iteration

----------------------------------------------------------------------
Ran 1997 tests in 82.103s

FAILED (errors=1, skipped=948)
Test failed: <unittest.runner.TextTestResult run=1997 errors=1 failures=0>
error: Test failed: <unittest.runner.TextTestResult run=1997 errors=1 failures=0>
Exception in thread Thread-11:
Traceback (most recent call last):
  File "/nix/store/pk2s4pk14hsvy410hcpc1hfscy34fn8p-python-2.7.15/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/nix/store/q2svfj4gqvz77mq7sf0mm796kbic580q-python2.7-redis-2.10.6/lib/python2.7/site-packages/redis/client.py", line 2599, in run
    timeout=sleep_time)
  File "/build/fakeredis-0.16.0/fakeredis.py", line 230, in decode_response
    val = _decode(func(*args, **kwargs))
  File "/build/fakeredis-0.16.0/fakeredis.py", line 508, in func_wrapper
    return func(*args, **kwargs)
  File "/build/fakeredis-0.16.0/fakeredis.py", line 2738, in get_message
    message = self._q.get(True, timeout)
  File "/nix/store/pk2s4pk14hsvy410hcpc1hfscy34fn8p-python-2.7.15/lib/python2.7/Queue.py", line 172, in get
    endtime = _time() + timeout
TypeError: 'NoneType' object is not callable
```

**Comment by bmerry** [comment]
I've been working on a rewrite of fakeredis, on a "rewrite" branch. Could you check if it still happens there? I think this might be because master locks at the wrong scope (separate lock per connection instead of on the shared state), and that should be fixed.

**Comment by bmerry** [comment]
I'm going to assume it's fixed in the rewrite branch (and 1.0rc1), but please re-open if you encounter more problems.
